src/server/protocol.ts

```diff
@@ -3017,6 +3029,39 @@ export interface LargeFileReferencedEventBody {
     maxFileSize: number;
 }
 
+export type CreateFileWatcherEventName = "createFileWatcher";

```

Comment:
Just writing this down before I forget (I haven't read the full PR yet), but one thing we may want to consider is preemptively modeling this after LSP, specifically, moving the glob calculation to the server and sending that as part of the request instead. Right now, that's done over on the VS Code side (per https://github.com/sheetalkamat/vscode/pull/2/files), but in the LSP, there is only one file watching API and it accepts globs (https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workspace_didChangeWatchedFiles).

This would mean that we would eventually have to pull the code back out of VS Code into tsserver anyway if we were to start using LSP.

Reply:
Also of note is that LSP doesn't have watch IDs or anything; when changing watches, you have to send the entire list of desired watches. That feels expensive but I'm not sure what all we watch (I have yet to view a trace of the current PR set to see).

Reply:
we are sending directory names to watch along with whether to watch recursively or not . I feel its better for API users to convert to whichever pattern they need it.

Ids are added because its easier and cheaper to figure out which Watcher to invoke instead of having to traffic around the complete data about paths and recursive etc info